trin-storage/src/versioned/store.rs

```diff
@@ -0,0 +1,107 @@
+use ethportal_api::{types::distance::Distance, OverlayContentKey};
+use trin_metrics::storage::StorageMetricsReporter;
+
+use crate::{
+    error::ContentStoreError, ContentId, ContentStore, PortalStorageConfig, ShouldWeStoreContent,
+};
+
+use super::{ContentType, StoreVersion};
+
+/// A trait for the versioned content store. Instance of it should be created using
+/// `create_store` function.
+pub trait VersionedContentStore: Sized {
+    /// Returns the version of the store.
+    fn version() -> StoreVersion;
+
+    /// Migrates content from previous version to the new version.
+    fn migrate_from(
+        content_type: ContentType,
+        old_version: StoreVersion,
+        config: &PortalStorageConfig,
+    ) -> Result<(), ContentStoreError>;
+
+    /// Creates the instance of the store. This shouldn't be used directly. Store should be
+    /// created using `create_store` function.
+    fn create(
```

Comment:
I don't like that anybody can just call this (as comments says, `create_store` function should be used). But I don't know how to enforce that.

I'm open for suggestions